Kernel/Devices/SerialDevice.cpp

```diff
@@ -140,4 +100,175 @@ u8 SerialDevice::get_line_status() const
     return m_base_addr.offset(5).in<u8>();
 }
 
+ssize_t SerialDevice::on_tty_write(const UserOrKernelBuffer& buffer, ssize_t size)
+{
+    if (!size)
+        return 0;
+
+    if (!(get_line_status() & EmptyTransmitterHoldingRegister))
+        return 0;
+
+    auto result = buffer.read_buffered<128>(size, [&](u8 const* data, size_t data_size) {
+        for (size_t i = 0; i < data_size; i++)
+            put_char(data[i]);
+        return data_size;
+    });
+    if (result.is_error())
+        return result.error();
+    return (ssize_t)result.value();
+}
+
+void SerialDevice::echo(u8 ch)
+{
+    auto buffer = UserOrKernelBuffer::for_kernel_buffer(&ch);
+    on_tty_write(buffer, 1);
+}
+
+void SerialDevice::flush_input()
+{
+    set_fifo_control(ClearReceiveFIFO);
```

Comment:
I'm pretty sure that this *clears* the buffers, not flushes it, aka it discards any characters not sent yet

Reply:
This is pretty weird, but both the Single UNIX Specification and the Linux manages say that these should discard pending IO. Will rename these to make it clear.

Reply:
theres a linux manpage about interacting with com ports?
(unelss you mean the linux man page talks about the flush_input method, in which case that very much might be the case, im talkign about the ClearReceiveFIFO bit, not the method itself)

Reply:
The [documentation](https://en.wikibooks.org/wiki/Serial_Programming/8250_UART_Programming#FIFO_Control_Register) I'm using says that setting these will automatically perform the flushes. And I'm following `tcflush()`'s behavior in the API.

Reply:
Hmm, from the link you set, that doesn't sound like flushing the data to me:
`Bits 1 and 2 are used to clear the internal FIFO buffers. This is useful when you are first starting up an application where you might want to clear out any data that may have been "left behind" by a previous piece of software using the UART, or if you want to reset a communications connection`
(in case we have different meanings for this terms, when i think of clear, i think of discarding the data, while flushing means sending all of the data we have queued over to the other side)

Reply:
Sorry for the misunderstanding. The spec uses these terms differently, and I tried to follow that. However I see that this is confusing, so I'll use function names like 'wait_until_transmission_completes()` and `discard_pending_{input, output}`.

For terminals, they take *flush* to mean discarding any data that has not been processed yet.
> Upon successful completion, tcflush() discards data written to the object referred to by fildes (an open file descriptor associated with a terminal) but not transmitted, or data received but not read

For what you are describing, they use the term *drain*:
> The tcdrain() function shall block until all output written to the object referred to by fildes is transmitted. The fildes argument is an open file descriptor associated with a terminal.

What I'm not sure about is what it means to flush the TTY upon receiving a character that should send a signal.